server/app/Http/Controllers/CommentController.php

```diff
@@ -6,137 +6,281 @@
+            $csv = $this->commentService->toCsv($comments);
+            $csv->output('comments.csv');
+            return;
+        } else {
+            $includeReplies = !$request->exists('include_replies') || $request->query('include_replies') && $request->query('include_replies') !== 'false';
+            $results = $comments->map(function ($item) use ($includeReplies) {
+                return $this->toAnnotatorArray($item, $includeReplies);
+            });
+
+            return Response::json($results);
         }
 
-        return Response::json($result);
     }
 
-    public function postIndex($doc)
+    public function getComment(DocAccessReadRequest $request, Doc $doc, Annotation $comment)
     {
-        $comment = Input::get('comment');
-
-        $newComment = new Comment();
-        $newComment->user_id = Auth::user()->id;
-        $newComment->doc_id = $comment['doc']['id'];
-        $newComment->text = $comment['text'];
-        $newComment->save();
+        if (!($comment->getRootTarget() instanceof Doc) || $comment->getRootTarget()->id != $doc->id) {
+            App::abort(404, 'A comment with id "'.$comment->id.'" does not exist on document with id "'.$doc->id);
+        }
 
-        // Late load the user.
-        $newComment->user;
+        return Response::json($this->toAnnotatorArray($comment));
+    }
 
-        return Response::json($newComment->toArray());
+    public function postIndex(DocAccessReadRequest $request, Doc $doc)
+    {
+        return $this->createComment($doc, Auth::user(), $request->all());
     }
 
-    public function postSeen($docId, $commentId)
+    public function postSeen(DocAccessEditRequest $request, Doc $doc, Annotation $comment)
     {
         $allowed = false;
 
         $user = Auth::user();
-        $user->load('docs');
 
-        // Check user documents against current document
-        foreach ($user->docs as $doc) {
-            if ($doc->id == $docId) {
-                $allowed = true;
-                break;
-            }
-        }
+        $this->annotationService->createAnnotationSeen($comment, $user, []);
 
-        if (!$allowed) {
-            throw new Exception("You are not authorized to mark this annotation as seen.");
-        }
+        Event::fire(new FeedbackSeen($comment, $user));
 
-        $comment = Comment::find($commentId);
-        $comment->seen = 1;
-        $comment->save();
+        return Response::json($this->toAnnotatorArray($comment));
+    }
 
-        Event::fire(new FeedbackSeen($comment, $user));
+    public function postLikes(DocAccessReadRequest $request, Doc $doc, Annotation $comment)
+    {
+        $this->annotationService->createAnnotationLike($comment, Auth::user(), []);
 
-        return Response::json($comment);
+        return Response::json($this->toAnnotatorArray($comment));
     }
 
-    public function postLikes($docId, $commentId)
+    public function postFlags(DocAccessReadRequest $request, Doc $doc, Annotation $comment)
```

Comment:
Need to have some business logic in here to prevent multiple flags from the same user. Perhaps in the FormRequest object, or in the Annotation service.